Diving in the pool

guilledk opened the pull request:
At the moment only pool funding mechanisms are implemented, also one test that covers the whole code except the `cancelfund` action. A simple direct funding is allowed now for testing, so anyone can alter pool funds under current implementation.

In the future we could use ballots to completely decentralize pool creation & funding:

1. First ballot to trigger pool creation, funding goals for initial commodity & currency reserves are set.
2. Once goals are set, a community crowd funding effort happens, were participants get a fee participation token.
3. Goals are reached, pool gets enabled to start making conversions.

Also a bug was found and fixed in the main `telosbookdex` contract, on the `withdraw` function, it was one of those pesky `contract` identifier collisions.

PD: Several fixes on `telosbookdex` testing APIs!

Commit:
Initial telospooldex commit

Commit:
Start using switch case to parse memo, disable inverted market creation and add memo == skip check to send tokens to contract

guilledk commented:
Last commit should close #56 and #55

Commit:
Add openpool.v1 protocol docs
Change inverse market creation error condition, now it lets it be created but only if normal doesn't exist, and then it blocks normal creation
Add openpool.v1 token conversion implementation to telospooldex
Add transfer echo contract
Drop koinonospool in favor of new generic telospookdex
Add very useful helpers to utils, asset_from_string, join, etc
Add tests for single conversion and multi step conversions

guilledk commented:
First version of decentralized token conversions done! ~~~~Wooooo

Viterbo commented on docs/openpool-proto.md:
````diff
@@ -0,0 +1,71 @@
+### Standard tables for `pools`
+
+```C++
+name symbols_get_index(symbol_code A, symbol_code B) {
````
It can fail when the symbol names are too large:
eg: symbols_get_index("TLOSDAC", "HEART") --> Fails

guilledk replied:
To fix this changed the return type of the function from `eosio::name` to `uint128_t`

Viterbo commented on contracts/telospooldex/telospooldex.cpp:
```diff
@@ -0,0 +1,297 @@
+                    get_contract_for_token(total.symbol.code()),
+                    "transfer"_n, 
+                    make_tuple(
+                        get_self(), recipient, total, string(""))
```
When sending the final tokens to the end-user after conversion, a "thank you" memo should be included for using the telos pool DEX. Something like: "Thank you for using Telos Pool DEX"

Viterbo commented on contracts/telospooldex/telospooldex.cpp:
```diff
@@ -0,0 +1,297 @@
+
+            vector<string> conversion_data = split(jumps.front(), "/");
+            check(conversion_data.size() == 2, ERR_MEMO_PARSING);
+            check(name(conversion_data[0]) == get_self(), ERR_INCORRECT_CONVERTER);
```
If the first converter in the path is not telospooldex then it fails.

Viterbo commented on contracts/telospooldex/telospooldex.cpp:
```diff
@@ -0,0 +1,297 @@
+            check(pool_it->commodity_reserve.amount > 0, ERR_POOL_NOT_FUNDED);
+            check(pool_it->currency_reserve.amount > 0, ERR_POOL_NOT_FUNDED);
+
+            // make conversion
```
After conversion, each converter should register a history entry in the correct market scope just like a normal deal for an order in the existing Telos (book) DEX:
https://github.com/guilledk/vapaee-smart-contracts/blob/telospooldex/include/vapaee/dex/modules/exchange.hpp#L515

Viterbo commented on contracts/telospooldex/telospooldex.hpp:
```diff
@@ -0,0 +1,131 @@
+
+#define PROTO_VERSION "openpool.v1"_n
+
+#define ECHO_CONTRACT "echocontract"_n
```
I already bought the name echocontract:
https://telos.bloks.io/account/echocontract

Commit:
Change openpool protocol function symbols_get_index return from eosio::name to uint128_t to fix issues with big token names.

Commit:
Change ``telosbookdex`` market indexing mechanisms drop ``table`` of type ``name`` in favour of ``uint128_t`` secondary index
Delete unused ``hotfix`` action from ``telosbookdex``
Add several util tests ``testcontract`` ```getindex, rawname, rawsymcode```
Simplify ``symbols_get_index`` logic
Some general cleanups pertaining to code style
Refactor some ``telosbookdex::market`` auxiliary functions
Fix minor bug on ballot/setcurrency tests

Commit:
Add gains > 0 check on main order processing code
Minor fix on dex::record module
Fix some record tests to find the right event entry
Move utils test to own subdirectory cause it was glitching pytest hard (performed two inits)

Viterbo commented on contracts/telospooldex/telospooldex.hpp:
```diff
@@ -0,0 +1,133 @@
+
+    namespace pool {
+
+        name get_contract_for_token(symbol_code sym) {
```
This function should be placed in include/vapaee/dex/modules/utils.hpp or include/vapaee/dex/modules/token.hpp (your choise)

Viterbo commented on contracts/telospooldex/telospooldex.hpp:
```diff
@@ -0,0 +1,133 @@
+            return book_it->contract;
+        }
+
+        asset get_exchange_rate(uint64_t pool_id) {
```
this function should be named **get_pool_rate** instead of **get_exchange_rate**

Viterbo commented on contracts/telospooldex/telospooldex.hpp:
```diff
@@ -0,0 +1,133 @@
+                using contract::contract;
+
+                [[eosio::action]]
+                void initpool(name creator, uint64_t market_id);
```
this action should be renamed to **createpool** because **initpool** may be confused with the contract initialization

Viterbo commented on contracts/telospooldex/telospooldex.hpp:
```diff
@@ -0,0 +1,133 @@
+                void initpool(name creator, uint64_t market_id);
+
+                [[eosio::action]]
+                void tryfund(
```
this action should be renamed to **startfund** because **tryfund** gives the notion of actually trying to fund (with actual funds, when there are no funds at this point). This is the start of a funding process, so the name should reflect that.

Viterbo commented on contracts/telospooldex/telospooldex.cpp:
```diff
@@ -0,0 +1,297 @@
+    auto commodity_it = book_tokens.find(book_it->commodity.raw());
+    auto currency_it = book_tokens.find(book_it->currency.raw());
+    check(commodity_it != book_tokens.end(), ERR_TOKEN_NOT_REG);
+    check(currency_it != book_tokens.end(), ERR_TOKEN_NOT_REG);
```
All this code is not needed. it can be deleted. Use the function **get_contract_for_token**() instead

Viterbo commented on contracts/telospooldex/telospooldex.cpp:
```diff
@@ -0,0 +1,297 @@
+    if (fund_it->commodity.amount > 0)
+        action(
+            permission_level{get_self(), "active"_n},
+            commodity_it->contract,
```
Use the function **get_contract_for_token**() instead

Viterbo commented on contracts/telospooldex/telospooldex.cpp:
```diff
@@ -0,0 +1,297 @@
+    if (fund_it->currency.amount > 0)
+        action(
+            permission_level{get_self(), "active"_n},
+            currency_it->contract,
```
Use the function **get_contract_for_token**() instead

Viterbo commented on contracts/telospooldex/telospooldex.cpp:
```diff
@@ -0,0 +1,297 @@
+
+        case "directfund"_n.value : {
+            // TODO: this is testing code to do initial funding 
+            uint64_t pool_id = strtoull(memo_tokens[1].c_str(), nullptr, 10);
```
need to check if we actually have memo_tokens[1] before using it
Use a specific error message for failing parsing a "directfund" attempt
`check(memo_tokens.size() > 1, ERR_MEMO_PARSING_DIRECTFUND);`

Viterbo commented on contracts/telospooldex/telospooldex.cpp:
```diff
@@ -0,0 +1,297 @@
+        }
+
+        case "fund"_n.value : {
+            uint64_t attempt_id = strtoull(memo_tokens[1].c_str(), nullptr, 10);
```
need to check if we actually have memo_tokens[1] before using it
Use a specific error message for failing parsing a "fund" attempt
`check(memo_tokens.size() > 1, ERR_MEMO_PARSING_FUND);`

Viterbo commented on contracts/telospooldex/telospooldex.cpp:
```diff
@@ -0,0 +1,297 @@
+        }
+
+        case PROTO_VERSION.value : {  
+            // general protocol parsing
```
need to check if we actually have memo_tokens[2] and memo_tokens[3] before using them
Again, use a specific error message.

Viterbo commented on contracts/telospooldex/telospooldex.cpp:
```diff
@@ -0,0 +1,297 @@
+            if (quantity.symbol == fund_it->currency.symbol)
+                funding_attempts.modify(fund_it, from, [&](auto &row) {
+                    row.currency += quantity;
+                });
```
Use **get_first_receiver()** to get the **token_contract** used in the transfer action being handled.
Then make sure the transferred funds are **not fake** comparing the token_contract with the commodity and currency contracts.

Fail with a proper message if the transferred funds are not one of the market tokens.

Viterbo commented on contracts/telospooldex/telospooldex.cpp:
```diff
@@ -0,0 +1,297 @@
+                });
+
+                funding_attempts.erase(fund_it);
+            }
```
This code should be a separated action called from outside named **endfund**. The typical use of the whole prodecure of funding should be a unique transaction including 4 actions: **startfund**, **transfer** A, **transfer** B, **endfund**

(to be continued...)

Commit:
Merge pull request #2 from guilledk/market_indexing

Change telosbookdex market indexing mechanisms

Commit:
Fix most of the low hanging fruit @viterbo pointed out in the latest review

Commit:
Change fund attempt indexing mechanism
Drop tryfund and simplify pool funding protocol according to @Viterbo's review
Move pool helper functions to own utils module

Viterbo commented on contracts/telospooldex/telospooldex.hpp:
```diff
@@ -0,0 +1,36 @@
+                using contract::contract;
+
+                [[eosio::action]]
+                void createpool(name creator, uint64_t market_id);
```
**createpool** must not be an external action. The pool must be created at the same time it is funded because it cannot exist without liquidity.

Viterbo commented on include/vapaee/pool/utils.hpp:
```diff
@@ -0,0 +1,149 @@
+                });
+            }
+
+            void maybe_end_fund_attempt(name funder, uint64_t market_id) {
+                fund_attempts funding_attempts(contract, contract.value);
+                auto pack_index = funding_attempts.get_index<"idpacked"_n>();
+                auto fund_it = pack_index.find(pack(funder.value, market_id));
+                
+                check(fund_it != pack_index.end(), ERR_ATTEMPT_NOT_FOUND);
+
+                asset commodity_ex = asset_change_precision(
+                        fund_it->commodity, ARITHMETIC_PRECISION);
+                asset currency_ex = asset_change_precision(
+                        fund_it->currency, ARITHMETIC_PRECISION);
+
+                asset fund_rate = asset_divide(commodity_ex, currency_ex);
+                
+                if (fund_rate == get_pool_rate(fund_it->get_pool_id())) {
+                    // TODO: emit fee participation token
+                    
+                    pools pool_markets(contract, contract.value);
+                    auto pool_it = pool_markets.find(fund_it->get_pool_id());
+                    check(pool_it != pool_markets.end(), ERR_POOL_NOT_FOUND);
+
+                    pool_markets.modify(pool_it, contract, [&](auto &row) {
+                        row.commodity_reserve += fund_it->commodity;
+                        row.currency_reserve += fund_it->currency;
+                    });
+
+                    pack_index.erase(fund_it);
+                }
+            }
```
Turn this "maybe" into an "always" (not talking about the title but the logic inside this function).

Pseudocode for this function:
```c++
if (exist_pool_for_market(market_id)) {

    // take the current ratio for the pool
    pool_rate = get_pool_rate(fund_it->get_pool_id());
    fund_rate = get_fund_rate(funt_it);
    current_liquidity = get_current_liquidity_total_credits(market_id);

    if (pool_rate > fund_rate) {
        // There's a surplus of currency in fund attempt
        diff = calculate_surplus_of_currency(pool_rate, fund_it);
        fund_pool(market_id, fund_it->commodity);
        fund_pool(market_id, fund_it->currency - diff);
        // return surplus tokens to funder
        transfer_tokens(contract, funder, diff, "Surplus in funding attempt to market " + market.toString());

    } else if (pool_rate < fund_rate) {
        // There's a surplus of commodity in fund attempt
        diff = calculate_surplus_of_commodity(pool_rate, fund_it);
        fund_pool(market_id, fund_it->commodity - diff);
        fund_pool(market_id, fund_it->currency);
        // return surplus tokens to funder
        transfer_tokens(contract, funder, diff, "Surplus in funding attempt to market " + market.toString());

    } else {
        // use the total amount of tokens provided for the fund attempt
        fund_pool(market_id, fund_it->commodity);
        fund_pool(market_id, fund_it->currency);
    }

    // gives the investor a number of credits
    // equivalent to the percentage of the liquidity provided
    // relative to the existing pool liquidity    
    liquidity_provided = get_current_liquidity_total_credits(market_id) - current_liquidity;
    creadits_for_investor(funder, market_id, liquidity_provided);

} else {
    // pool does not exist. So we created now
    create_pool_for_market(market_id);

    // use the total amount of tokens provided for the fund attempt
    fund_pool(market_id, fund_it->commodity);
    fund_pool(market_id, fund_it->currency);

    // gives the investor a number of credits
    // equivalent to the 100% of the liquidity of this pool
    total_liquidity = get_current_liquidity_total_credits(market_id);
    creadits_for_investor(funder, market_id, total_liquidity);
}

```

Commit:
Per @Viterbo's request, add imprecise funding attempt logic, and surplus return machinery
Change name & logic of ``maybe_end_fund_attempt` to ``end_fund_attempt`` (always ends fund attempt)
Moved create pool logic to `telospooldex` utils module
Remove createpool action, pools are created on demand using `directfund`
Add pool exist check on `fund`
Add ``asset_change_symbol`` helper to base utils, also renamed some vars

guilledk commented:
TODO:

- [x] Add tests to return surplus on fund machinery
- [x] Add conversion & fund attempt history 
- [x] Add participation token logic

Commit:
Add tests to surplus cases on funding attempt handler

Commit:
Add conversion history

Commit:
Add simple conversion history tests, refactor telosbookdex constants file to use --upper on get table

Commit:
Add fund attempt history & tests
Also add comments to tests

Commit:
Update test APIs, and change default test sym precision from 2 to 4 to avoid invert errors.

Viterbo commented on include/vapaee/base/utils.hpp:
```diff
@@ -121,6 +120,30 @@ namespace vapaee {
+        }
+
         int128_t divide(const asset &A, const asset &B) {
             check(A.symbol.precision() == B.symbol.precision(),
```
check B.amount > 0

Viterbo commented on contracts/telospooldex/telospooldex.cpp:
```diff
@@ -0,0 +1,256 @@
+
+        case "fund"_n.value : {
+            check(memo_tokens.size() == 2, ERR_MEMO_PARSING);
+            uint64_t market_id = strtoull(memo_tokens[1].c_str(), nullptr, 10);
```
Allow the protocol to support both the market_id and the symbol pair (commodity/currency). That couple will have a canonical representation for which we already have a function. Use that representation to find the market_id;

both memos should be valid:
"fund,0" & "fund,CNT/TLOS"

Viterbo commented on contracts/telospooldex/telospooldex.cpp:
```diff
@@ -0,0 +1,256 @@
+            
+            pools pool_markets(get_self(), get_self().value);
+            auto pool_it = pool_markets.find(market_id);
+            check(pool_it != pool_markets.end(), ERR_POOL_NOT_FOUND);
```
Allow starting funding for a non-existing-yet pool and postpone the pool creation until its full funding is imminent (end_fund_attempt). In other words, delete these lines.

Viterbo commented on include/vapaee/pool/utils.hpp:
```diff
@@ -0,0 +1,270 @@
+                });
+            }
+
+            void end_fund_attempt(name funder, uint64_t market_id) {
```
Implement the if structure with the missing part of the pseudo code that is here: https://github.com/vapaee/vapaee-smart-contracts/pull/53#discussion_r699299298
```c++
if (exist_pool_for_market (market_id)) {
    CODE_ALREADY_DONE_OK;
} else {
    MISSING_POOL_CREATION_PART;
}   
```

This is related to the deletion of the three lines (105-107) in file https://github.com/guilledk/vapaee-smart-contracts/blob/telospooldex/contracts/telospooldex/telospooldex.cpp#L105

Viterbo commented on contracts/telospooldex/telospooldex.cpp:
```diff
@@ -0,0 +1,256 @@
+void telospooldex::cancelfund(name funder, uint64_t market_id) {
+    require_auth(funder);
+
+    fund_attempts funding_attempts(get_self(), get_self().value);
+    auto pack_index = funding_attempts.get_index<"idpacked"_n>();
+    auto fund_it = pack_index.find(pack(funder.value, market_id));
```
fund_attempts should use the funder as scope and the market_id as primary key:

```c++
fund_attempts funding_attempts(get_self(), funder.value);
auto fund_it = funding_attempts..find(market_id);
```

Viterbo commented on include/vapaee/pool/utils.hpp:
```diff
@@ -0,0 +1,270 @@
+            }
+
+            void end_fund_attempt(name funder, uint64_t market_id) {
+                fund_attempts funding_attempts(contract, contract.value);
+                auto pack_index = funding_attempts.get_index<"idpacked"_n>();
+                auto fund_it = pack_index.find(pack(funder.value, market_id));
```
fund_attempts should use the funder as scope and the market_id as primary key:

```c++
fund_attempts funding_attempts(get_self(), funder.value);
auto fund_it = funding_attempts..find(market_id);
```

Commit:
Simplify fund_attempt table indexing mechanism
Add support for addressing markets by their symbol pair in fund attempt, with simple test
Add divide by zero warning on utils::divide

Commit:
Refactor fund mechanic to create the pool on demand and send initial funds
Remove directfund mechanic

Add test to show market indexing bug, also add some debug prints

Commit:
Add comments to bug demonstration test

Viterbo commented on tests/telosbookdex/test_market_ids.py:
```diff
@@ -0,0 +1,102 @@
+    divisible by two.
+
+    I think the problem is in:
+        ``vapaee::dex::market::aux_get_canonical_index_for_symbols``
```
Yes, It was. You found an old bug thanks to this test!! Good job!

The problem is in [line 110 of market](https://github.com/guilledk/vapaee-smart-contracts/blob/c1ce2955aa859ab9f9da3fd38c04c5b264574cc8/include/vapaee/dex/modules/market.hpp#L110) module. The second "if" should be the opposite of the first but instead it's a copy-paste! So, it never enters in the case of B being currency for A in a secondary group.
```c++
// ...
else if (!A_is_currency_for_B && B_is_currency_for_A)
// ...
```

Commit:
Add @Viterbo's fix to intermitent cannonical index bug!
Finish removeing old direct fund mechanism

Commit:
Refactor history records to make table identical to telosbookdex format

Commit:
Refactor pool conversion history to use telosbookdex aux_register_transaction_in_history

Commit:
Return funding history to its initial form

Commit:
Add inital funder reward system, begin tracking participation score for every funding event

Commit:
Add takepart action to allow funders to withdraw their investments

Commit:
Add config singleton to pooldex
Add conversion fee
Add pool roi test

Viterbo committed:
Some clean up

Viterbo committed:
Refactoring code - table names and filenames should be the same

Viterbo committed:
ACTION newmarket

Commit:
Merge pull request #4 from Viterbo/telospooldex

Minor changes on telospooldex branch
